### PR TITLE
Fixes #20310: Method getByRulesCompliance used by API is highly inefficient

### PR DESCRIPTION
**Pull request by ncharles**
https://issues.rudder.io/issues/20310

### CODE REVIEW

**Comment by ncharles** [comment]
This PR is not finished, but so far:
* API calls to compliance/rules go from 17s to 14s (it returns 229MB of data)
* it changes what is returned however: Rules not applied on anything don't show up in the resulting list - i'm unsure whether it's good or not, at least for me it seems to be a change that make sense (no compliance is not 0 compliance)

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
This PR changes what is returned:  Rules not applied on anything don't show up in the resulting list - i'm unsure whether it's good or not, at least for me it seems to be a change that make sense (no compliance is not 0 compliance)

API calls to compliance/rules go from 17s to 14s with 4950 nodes (include js serialization)
Without writing json output:
API calls for level1 goes from 5s to 1s
API calls for level2 goes from 5s to 0.6s
API calls for level3 stays at 5.6s

**Comment by ncharles** [comment]
PR updated with a new commit

**Comment by ncharles** [comment]
actually, we should not change the list of rules returns, as it is used in 7.0 for rule pages

**Comment by ncharles** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 1]
```diff
@@ -98,7 +97,12 @@ class ComplianceApi(
 
       (for {
         level <- restExtractor.extractComplianceLevel(req.params)
-        rules <- complianceService.getRulesCompliance()
+        computeLevel <- Full(if(version.value <= 6) {
+                          None
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:101]: Should be Some(10)  (or Some(value expected for api versio 6 )

But Version 6 is not supported anymore with we just may want to remove them (need to check though), but i think 6.1 is 7 minimal api version

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:101]: a few line below you can use compluteLevel insted of level.getOrElse(10)

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:101]: Also this fails if level is not defined, if defined and version > 6 we should default to default vlaue (10)

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:101]: Ok I misread i haven't seen the Full, nevermind this is ok !

Comment (ncharles) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:101]: level is an option that appears only later, so I'd like to keep the semantic of "not defined" , thus an  None

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:101]: Yes i just misread and haven't seen it was in a Full and thought the None could break the computation, but it's not the case at all

**Comment by ncharles** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 2]
```diff
@@ -232,109 +235,161 @@ class ComplianceAPIService(
...
- private[this] def getByRulesCompliance(rules: Set[Rule]) : Box[Seq[ByRuleRuleCompliance]] = {
-
+ private[this] def getByRulesCompliance(rules: Seq[Rule], level: Option[Int]) : Box[Seq[ByRuleRuleCompliance]] = {
+    val computedLevel = level.getOrElse(10)
+    val t1 = System.currentTimeMillis()
     for {
       groupLib      <- nodeGroupRepo.getFullGroupLibrary().toBox
-      directivelib  <- directiveRepo.getFullDirectiveLibrary().toBox
+      t2             = System.currentTimeMillis()
+      _              = logger.trace(s"getByRulesCompliance - getFullGroupLibrary in ${t2 - t1} ms")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:254]: ~400ms

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 3]
```diff
@@ -232,109 +235,161 @@ class ComplianceAPIService(
...
+                         directiveRepo.getFullDirectiveLibrary().toBox.map(_ . allDirectives )
+                       } else {
+                         Full(Map[DirectiveId, (FullActiveTechnique, Directive)]())
+                       }
+      t3             = System.currentTimeMillis()
+      _              = logger.trace(s"getByRulesCompliance - getFullDirectiveLibrary in ${t3 - t2} ms")
+
       nodeInfos     <- nodeInfoService.getAll()
+      t4             = System.currentTimeMillis()
+      _              = logger.trace(s"getByRulesCompliance - nodeInfoService.getAll() in ${t4 - t3} ms")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:267]: ~ 8ms

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 4]
```diff
@@ -232,109 +235,161 @@ class ComplianceAPIService(
...
+
+
+      ruleObjects   = rules.map { case x => (x.id, x) }.toMap
       reportsByNode <- reportingService.findRuleNodeStatusReports(
-                         nodeInfos.keySet, rules.map(_.id).toSet
+                         nodeInfos.keySet, ruleObjects.keySet
                        )
-    } yield {
+      t6             = System.currentTimeMillis()
+      _              = logger.trace(s"getByRulesCompliance - findRuleNodeStatusReports in ${t6 - t5} ms")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:279]: ~ 170ms

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 5]
```diff
@@ -232,109 +235,161 @@ class ComplianceAPIService(
...
-            , rule.name
-            , ComplianceLevel(noAnswer = nodeIds.size)
-            , compliance.mode
-            , Seq()
-          ))
-        }).toMap
-      }
+      val reportsByRule  = reportsByNode.flatMap { case(_, status) => status.reports }.groupBy( _.ruleId)
+      val t7             = System.currentTimeMillis()
+      logger.trace(s"getByRulesCompliance - group reports by rules in ${t7 - t6} ms")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:284]: ~150 ms

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 6]
```diff
@@ -232,109 +235,161 @@ class ComplianceAPIService(
...
-          )
-        )
-      }.toMap
 
-      //return the full list, even for non responding nodes/directives
-      //but override with values when available.
-      (initializedCompliances ++ nonEmptyRules).values.toSeq
+              }
+      val t8        = System.currentTimeMillis()
+      logger.trace(s"getByRulesCompliance - Compute non empty rules in ${t8 - t7} ms")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:340]: ~120ms

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 7]
```diff
@@ -232,109 +235,161 @@ class ComplianceAPIService(
...
+                , ComplianceLevel(noAnswer = nodeIds.size)
+                , compliance.mode
+                , Seq()
+              )
+            }
+          }
+        }
+
+      val t9 = System.currentTimeMillis()
+      logger.trace(s"getByRulesCompliance - Compute ${initializedCompliances.size} empty rules in ${t9 - t8} ms")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:368]: ~ 60ms

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala (hunk 2 above)
Comment (ncharles) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:254]: timings for 4950 nodes, with API answering in 1.2s

**Review by VinceMacBuche — APPROVED** [review]

**Comment by ncharles** [comment]
OK, squash merging this PR

**Comment by ncharles** [comment]
OK, squash merging this PR

**Comment by ncharles** [comment]
OK, squash merging this PR

**Comment by amousset** [comment]
OK, squash merging this PR

**Comment by amousset** [comment]
OK, squash merging this PR

**Comment by ncharles** [comment]
OK, squash merging this PR

**Comment by ncharles** [comment]
OK, squash merging this PR

**Comment by ncharles** [comment]
OK, squash merging this PR

**Comment by ncharles** [comment]
OK, squash merging this PR

**Comment by ncharles** [comment]
OK, merging this PR

**Comment by ncharles** [comment]
OK, merging this PR

**Pull request merged**